Add new lint "missing_panics_doc"

brightly-salty opened the pull request:
fixes #1974 
changelog: Added the "missing_panics_doc" lint which lints when public functions that may panic are missing "# Panics" in their doc comment

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @flip1995 (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust-clippy/blob/master/CONTRIBUTING.md) for more information.

flip1995 commented on clippy_lints/src/doc.rs:
```diff
@@ -609,3 +694,43 @@ fn check_word(cx: &LateContext<'_>, word: &str, span: Span) {
+    }
+
+    fn nested_visit_map(&mut self) -> NestedVisitorMap<Self::Map> {
+        NestedVisitorMap::None
```
You want to at least have [`NestedVisitorMap::OnlyBodies`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/enum.NestedVisitorMap.html#variant.OnlyBodies) here, so that closures inside functions get visited.

Otherwise 

```rust
fn foo(opt: Option<u32>) {
    opt.map(|x| if x == 10 { panic!() });
}
```
won't get linted.

Please also add a test for this.

---

The question is, if you even want to use `NestedVisitorMap::All`, which would also lint inner functions. One argument in favor would be, that inner functions are called somewhere in the body (Otherwise rustc would emit an `unused` warning). The argument against this is to be conservative and have this as a potential FN.

brightly-salty replied:
(I just copied and pasted this code, with a few adjustments, from one of the lints that found `todo!()` and things like that.) I'll do which ever one you thing is best.

flip1995 replied:
In that case, just use `OnlyBodies`, but also add a test for a nested function, that might panic.

brightly-salty replied:
I suppose I should do the following?
> If for some reason you want the nested behavior, but don't have a Map at your disposal: then you should override the visit_nested_XXX methods, and override this method to panic!(). This way, if a new visit_nested_XXX variant is added in the future, we will see the panic in your code and fix it appropriately.
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/trait.Visitor.html#tymethod.nested_visit_map

flip1995 replied:
No you have a HIR map, but it is hidden in the `TyCtxt` of the `LateContext`.

```suggestion
        NestedVisitorMap::OnlyBodies(self.cx.tcx.hir())
```

flip1995 reviewed (changes requested):
Welcome to Clippy! :heart: 

This looks really good to me!

flip1995 commented on clippy_lints/src/doc.rs:
```diff
@@ -609,3 +694,43 @@ fn check_word(cx: &LateContext<'_>, word: &str, span: Span) {
 }
+
+struct FindPanicUnwrap<'a, 'tcx> {
+    lcx: &'a LateContext<'tcx>,
```
NIT: we usually abbreviate both `LateContext` and `EarlyContext` with just `cx` in Clippy.
```suggestion
    cx: &'a LateContext<'tcx>,
```

flip1995 commented on clippy_lints/src/doc.rs:
```diff
@@ -609,3 +694,43 @@ fn check_word(cx: &LateContext<'_>, word: &str, span: Span) {
+
+struct FindPanicUnwrap<'a, 'tcx> {
+    lcx: &'a LateContext<'tcx>,
+    typeck_results: &'tcx ty::TypeckResults<'tcx>,
```
Since you pass the `LateContext` anyway, this field shouldn't be necessary. Calling `cx.tcx.expr_ty(..)` is also more stable.

brightly-salty replied:
The problem is that we get the typecheck results form `cx.tcx.typeck(impl_item_def_id)` and otherwise we would have to have a def_id field

flip1995 reviewed (approved):
There is still a dogfood error. Otherwise LGTM

brightly-salty commented:
@flip1995 I fixed it, but now there's a merge conflict that I don't have the capability to resolve. Looks like it's just a changed function signature.

flip1995 commented:
I rebased it and added the first panic location in the lint output.

With this, I now understand why the `proc_macro` is linted. This is a FP, I'd like to see fixed, before merging this.

brightly-salty commented:
> I rebased it and added the first panic location in the lint output.
> 
> With this, I now understand why the `proc_macro` is linted. This is a FP, I'd like to see fixed, before merging this.

I don't think I understand. Your changes look good to me, just need to remove an unused import and add the span changes in another part of the codebase.

flip1995 commented on mini-macro/src/lib.rs:
```diff
@@ -7,6 +7,9 @@ extern crate proc_macro;
 use proc_macro::{quote, TokenStream};
 
 #[proc_macro_derive(ClippyMiniMacroTest)]
+/// # Panics
+///
+/// Panics if the macro derivation fails
```
This should not be necessary.  The panic happens in an external macro ``quote!`` which will be confusing to the user, if they get a warning from this lint.

brightly-salty replied:
That makes sense. Does that mean we need to change something in FindPanicUnwrap to not check if we are in a macro?

flip1995 replied:
Well we need to check somehow for macros in this visitor. The problem here is, that `in_external_macro` won't really work here, because ``panic`` and so on are also external macros (I think). It would be great, if you could try experimenting with `utils::is_expn_of` and `is_direct_expn_of` and then only lint when we're inside a ``std`` panic macro (``panic``/``unreachable``). Also maybe not ``todo`` or ``unimplemented``, since those macros signal that the implementation is unfinished and panics may not occur in the finished function. 

So rather than checking just for ``match_panic_def_id``, you can try to also check if the expansion is coming from ``panic`` or ``unreachable``.

brightly-salty replied:
Actually, after thinking it over some more, I think I respectfully disagree @flip1995 . Users should be aware of macros that panic within their function and document that. If we need to make the warning clearer to show that the macro panics, or point to the place in the macro where it panics, that's okay.

flip1995 replied:
That is a good point. And since this lint is `pedantic` I think this is reasonable. I'm convinced :+1:

flip1995 commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit bde667af7e7d512978daff3bc2b540bb913bd6a1 has been approved by `flip1995`

<!-- @bors r=flip1995 bde667af7e7d512978daff3bc2b540bb913bd6a1 -->
<!-- homu: {"type":"Approved","sha":"bde667af7e7d512978daff3bc2b540bb913bd6a1","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit bde667af7e7d512978daff3bc2b540bb913bd6a1 with merge 28794e956e2c344094d0dab1f1551eac9cb2f6d6...
<!-- homu: {"type":"BuildStarted","head_sha":"bde667af7e7d512978daff3bc2b540bb913bd6a1","merge_sha":"28794e956e2c344094d0dab1f1551eac9cb2f6d6"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/1815744505), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/1815738561), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/1815807879)
Approved by: flip1995
Pushing 28794e956e2c344094d0dab1f1551eac9cb2f6d6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/1815807879","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/1815744505","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/1815738561"},"merge_sha":"28794e956e2c344094d0dab1f1551eac9cb2f6d6"} -->
